src/Limiters/vertical_mass_borrowing_limiter.jl

```diff
@@ -0,0 +1,140 @@
+end
+# TODO: should q_min be a field that varies in space?
+
+
+"""
+    apply_limiter!(q::Fields.Field, ρ::Fields.Field, lim::VerticalMassBorrowingLimiter)
+
+Apply the vertical mass borrowing
+limiter `lim` to field `q`, given
+density `ρ`.
+"""
+apply_limiter!(
+    q::Fields.Field,
+    ρ::Fields.Field,
+    lim::VerticalMassBorrowingLimiter,
+) = apply_limiter!(q, ρ, axes(q), lim, ClimaComms.device(axes(q)))
+
+function apply_limiter!(
+    q::Fields.Field,
+    ρ::Fields.Field,
+    space::Spaces.FiniteDifferenceSpace,
+    lim::VerticalMassBorrowingLimiter,
+    device::ClimaComms.AbstractCPUDevice,
+)
+    cache = (; bmass = lim.bmass, ic = lim.ic, q_min = lim.q_min)
+    columnwise_massborrow_cpu(q, ρ, cache)
+end
+
+function apply_limiter!(
+    q::Fields.Field,
+    ρ::Fields.Field,
+    space::Spaces.ExtrudedFiniteDifferenceSpace,
+    lim::VerticalMassBorrowingLimiter,
+    device::ClimaComms.AbstractCPUDevice,
+)
+    Fields.bycolumn(axes(q)) do colidx
+        cache = (;
+            bmass = lim.bmass[colidx],
+            ic = lim.ic[colidx],
+            q_min = lim.q_min,
+        )
+        columnwise_massborrow_cpu(q[colidx], ρ[colidx], cache)
+    end
+end
+
+# TODO: can we improve the performance?
+# `bycolumn` on the CPU may be better here since we could multithread it.
+function columnwise_massborrow_cpu(q::Fields.Field, ρ::Fields.Field, cache) # column fields
+    # TODO: maybe rm some stuff from struct or cache
+    (; bmass, ic, q_min) = cache
+    Δz = Fields.Δz_field(q)
+    Δz_vals = Fields.field_values(Δz)
+    (; J) = Fields.local_geometry_field(ρ)
+    # ΔV_vals = Fields.field_values(J)
+    ΔV_vals = Δz_vals
+    ρ_vals = Fields.field_values(ρ)
+    #  loop over tracers
+    nlevels = Spaces.nlevels(axes(q))
+    @. ic = 0
+    @. bmass = 0
+    # TODO: Maybe delete ic
+    # TODO: make bmass positive
+    q_vals = Fields.field_values(q)
+    # top to bottom
+    for f in 1:DataLayouts.ncomponents(q_vals)
+        for v in nlevels:-1:1
+            CI = CartesianIndex(1, 1, f, v, 1)
+            # new mass in the current layer
+            ρΔV_lev =
+                DL.getindex_field(ΔV_vals, CI) * DL.getindex_field(ρ_vals, CI)
+            nmass = DL.getindex_field(q_vals, CI) + bmass[] / ρΔV_lev
+            # TODO: should this be \geq
+            if nmass > q_min[f]
+                #  if new mass in the current layer is positive, don't borrow mass any more
+                DL.setindex_field!(q_vals, nmass, CI)
+                bmass[] = 0
+            else
+                #  set mass to q_min in the current layer, and save bmass
+                bmass[] = (nmass - q_min[f]) * ρΔV_lev
+                DL.setindex_field!(q_vals, q_min[f], CI)
+                ic[] = ic[] + 1
+            end
+        end
+
+        #  bottom to top
+        for v in 1:nlevels
+            CI = CartesianIndex(1, 1, f, v, 1)
+            # if the surface layer still needs to borrow mass
+            if bmass[] < 0
```

Comment:
If the surface/bottom layer for a given column doesn't have negative tracer mass after the first top-to-bottom pass, the loop over `1:nlevels` is unnecessary right ? Is there a need to check this condition at every single level `v` ?